java/ql/src/semmle/code/java/security/UnsafeAndroidAccess.qll

```diff
@@ -0,0 +1,92 @@
+/**
+ * Provides classes to reason about Unsafe Resource Fetching vulnerabilities in Android.
+ */
+
+import java
+private import semmle.code.java.frameworks.android.WebView
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.dataflow.ExternalFlow
+
+/**
+ * A sink that represents a method that fetches a web resource in Android.
+ *
+ * Extend this class to add your own Unsafe Resource Fetching sinks.
+ */
+abstract class UrlResourceSink extends DataFlow::Node {
+  /**
+   * Returns a description of this vulnerability,
+   */
+  abstract string getSinkType();
+}
+
+/**
+ * Cross-origin access enabled resource fetch.
+ *
+ * It requires JavaScript to be enabled too to be considered a valid sink.
+ */
+private class CrossOriginUrlResourceSink extends JavaScriptEnabledUrlResourceSink {
+  CrossOriginUrlResourceSink() {
+    exists(MethodAccess ma, MethodAccess getSettingsMa |
+      ma.getMethod() instanceof CrossOriginAccessMethod and
+      ma.getArgument(0).(BooleanLiteral).getBooleanValue() = true and
+      ma.getQualifier().(VarAccess).getVariable().getAnAssignedValue() = getSettingsMa and
+      getSettingsMa.getMethod() instanceof WebViewGetSettingsMethod and
+      getSettingsMa.getQualifier().(VarAccess).getVariable().getAnAccess() =
+        this.asExpr().(Argument).getCall().getQualifier()
+    )
+  }
+
+  override string getSinkType() {
+    result = "user input vulnerable to cross-origin and sensitive resource disclosure attacks"
+  }
+}
+
+/**
+ * JavaScript enabled resource fetch.
+ */
+private class JavaScriptEnabledUrlResourceSink extends UrlResourceSink {
+  JavaScriptEnabledUrlResourceSink() {
+    exists(MethodAccess loadUrl, VarAccess webviewVa, MethodAccess getSettingsMa, Variable v |
+      loadUrl.getArgument(0) = this.asExpr() and
+      loadUrl.getMethod() instanceof WebViewLoadUrlMethod and
+      loadUrl.getQualifier() = webviewVa and
+      getSettingsMa.getMethod() instanceof WebViewGetSettingsMethod and
+      webviewVa.getVariable().getAnAccess() = getSettingsMa.getQualifier() and
+      v.getAnAssignedValue() = getSettingsMa and
+      isJSEnabled(v)
+    )
```

Comment:
Parts of this logic is repeated in the other charpred above in `CrossOriginUrlResourceSink`. Pull out the shared bits in a predicate, e.g.:
```ql
/**
 * Holds if a `WebViewLoadUrlMethod` method is called with the given `urlArg` on a
 * WebView with settings stored in `settings`.
 */
private predicate webViewLoadUrl(Expr urlArg, Variable settings) {
  exists(MethodAccess loadUrl, Variable webview, MethodAccess getSettings |
    loadUrl.getArgument(0) = urlArg and
    loadUrl.getMethod() instanceof WebViewLoadUrlMethod and
    loadUrl.getQualifier() = webview.getAnAccess() and
    getSettings.getMethod() instanceof WebViewGetSettingsMethod and
    webview.getAnAccess() = getSettingsMa.getQualifier() and
    settings.getAnAssignedValue() = getSettings
  )
}
```
Then this charpred can be written as
```suggestion
    exists(Variable settings |
      webViewLoadUrl(this.asExpr(), settings) and
      isJSEnabled(settings)
    )
```
And the above charpred
```ql
  CrossOriginUrlResourceSink() {
    exists(Variable settings, MethodAccess ma, MethodAccess getSettingsMa |
      webViewLoadUrl(this.asExpr(), settings) and
      ma.getMethod() instanceof CrossOriginAccessMethod and
      ma.getArgument(0).(BooleanLiteral).getBooleanValue() = true and
      ma.getQualifier() = settings.getAnAccess()
    )
  }
```